Replicator/Puller.cc

```diff
@@ -276,15 +276,21 @@ namespace litecore { namespace repl {
         }
     }
 
-    // Callback from an IncomingRev when it's finished (either added to db, or failed)
+    // Called from an IncomingRev when it's finished (either added to db, or failed.)
+    // The IncomingRev will be processed later by _revsFinished().
     void Puller::revWasHandled(IncomingRev *inc) {
-        _incomingDocIDs.remove(inc->rev()->docID);       // this is thread-safe
-        _returningRevs.push(inc);
+        // CAUTION: For performance reasons this method is called directly, without going through the
+        // Actor event queue, so it runs on the IncomingRev's thread, NOT the Puller's! Thus, it needs
+        // to pay attention to thread-safety.
+        _incomingDocIDs.remove(inc->rev()->docID);      // this is thread-safe
+        _returningRevs.push(inc);                       // this is thread-safe
     }
 
     void Puller::_revsFinished(int gen) {
         auto revs = _returningRevs.pop(gen);
         for (IncomingRev *inc : *revs) {
+            // If it was provisionally inserted, it'll have called _revWasProvisionallyHandled
+            // already. If not, call that method now:
             if (!inc->wasProvisionallyInserted())
```

Comment:
does this mean, all the revisions, including the rejected ones(which are not inserted to db), will get a flag called `wasProvisionallyInserted` on it?

Reply:
got confused with the inserted name, which i assumed, will only get callbacks for those are inserted to the db. Since the rejected revisions are not inserted, i assume, we should probably get back a callback which is like `handled` and not inserted.

Reply:
"Provisionally inserted" means the revision's been written to the database but the enclosing transaction hasn't been committed yet. (Revs get added in bulk.) This is a boolean flag on the IncomingRev class used by the replicator. When a rev is provisionally inserted, `Puller::_revWasProvisionallyHandled` is called.

If a rev is not inserted, it still goes through `_revsFinished`, and the bookkeeping needs to happen then.

Reply:
yea. understood now.